src/pyfdb_bindings/bindings.cc

```diff
@@ -0,0 +1,383 @@
+#include <map>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+#include "eckit/config/YAMLConfiguration.h"
+#include "eckit/exception/Exceptions.h"
+#include "eckit/filesystem/URI.h"
+#include "eckit/io/DataHandle.h"
+#include "eckit/runtime/Main.h"
+#include "fdb5/api/FDB.h"
+#include "fdb5/api/helpers/ControlIterator.h"
+#include "fdb5/api/helpers/FDBToolRequest.h"
+#include "fdb5/api/helpers/ListElement.h"
+#include "fdb5/api/helpers/ListIterator.h"
+#include "fdb5/api/helpers/MoveIterator.h"
+#include "fdb5/api/helpers/StatsIterator.h"
+#include "fdb5/config/Config.h"
+#include "fdb5/database/BaseKey.h"
+#include "fdb5/database/FieldLocation.h"
+#include "fdb5/database/IndexAxis.h"
+#include "fdb5/database/IndexStats.h"
+#include "fdb5/database/Key.h"
+#include "metkit/mars/MarsRequest.h"
+
+namespace py   = pybind11;
+namespace mars = metkit::mars;
+
+class PyDataHandle : public eckit::DataHandle, public py::trampoline_self_life_support {
+
+    /* Trampoline (need one for each virtual function) */
+    void print(std::ostream& s) const override {
+        PYBIND11_OVERRIDE_PURE(void,              /* Return type */
+                               eckit::DataHandle, /* Parent class */
+                               print,             /* Name of function in C++ (must match Python name) */
+        );
+    }
+    long read(void* buf, long length) override {
+        PYBIND11_OVERRIDE_PURE(long,              /* Return type */
+                               eckit::DataHandle, /* Parent class */
+                               read,              /* Name of function in C++ (must match Python name) */
+                               buf,               /* Name of parameter */
+                               length);           /* Name of parameter */
+    }
+};
+
+PYBIND11_MODULE(pyfdb_bindings, m) {
+    m.def("init_bindings", []() {
+        const char* args[] = {"pyfdb", ""};
+        eckit::Main::initialise(1, const_cast<char**>(args));
+    });
+
+    py::class_<fdb5::Config>(m, "Config")
+        .def(py::init([]() { return fdb5::Config(); }))
+        .def(py::init([](const std::string& config, const std::optional<std::string>& user_config) {
+            if (user_config.has_value()) {
+                return fdb5::Config(eckit::YAMLConfiguration(config), eckit::YAMLConfiguration(user_config.value()));
+            }
+            return fdb5::Config(eckit::YAMLConfiguration(config));
+        }))
+        .def("__str__", [](const fdb5::Config& config) {
+            std::stringstream buf;
+            buf << "Configuration: \n";
+            buf << config;
+            buf << "\nUser configuration: \n";
+            buf << config.userConfig();
+            return buf.str();
+        });
+
+    py::class_<eckit::DataHandle, PyDataHandle, py::smart_holder>(m, "DataHandle")
+        .def(py::init())
+        .def("open_for_read", [](eckit::DataHandle& data_handle) { data_handle.openForRead(); })
```

Comment:
This implementation is (essentially) a read data handle. I'm not convinced that we get value for explicitly having a function open_for_read in the absence of open_for_write and open_for append.